src/libAtomVM/nifs.c

```diff
@@ -2466,3 +2478,161 @@ static term nif_console_print(Context *ctx, int argc, term argv[])
     }
     return OK_ATOM;
 }
+
+static char b64_table[64] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+// per https://tools.ietf.org/rfc/rfc4648.txt
+
+static term nif_atomvm_b64_encode(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term src = argv[0];
+
+    VALIDATE_VALUE(src, term_is_binary);
+
+    size_t src_size = term_binary_size(src);
+    if (src_size == 0) {
+        return src;
+    }
+    uint8_t *src_buf = malloc(src_size);
+    if (IS_NULL_PTR(src_buf)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    memcpy(src_buf, term_binary_data(src), src_size);
+    uint8_t *src_pos = src_buf;
+    size_t dst_size = (4 * src_size) / 3;
+    size_t pad = 0;
+    switch (src_size % 3) {
+        case 0:
+            break;
+        case 1:
+            pad = 2;
+            dst_size++;
+            break;
+        case 2:
+            pad = 1;
+            dst_size++;
+            break;
+    }
+    if (UNLIKELY(memory_ensure_free(ctx, term_binary_data_size_in_terms(dst_size + pad) + BINARY_HEADER_SIZE) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term dst = term_create_empty_binary(dst_size + pad, ctx);
+    uint8_t *dst_pos = (uint8_t *) term_binary_data(dst);
+    for (size_t i = 0;  i < dst_size;  ++i) {
+        uint8_t accum = 0;
+        switch (i & 0x03) {
+            case 0:
+                dst_pos[i] = b64_table[(*src_pos) >> 2];
+                break;
+            case 1:
+                accum = ((*src_pos) & 0x03) << 4;
+                src_pos++;
+                if (i < dst_size - 1) {
+                    accum |= ((*src_pos) & 0xF0) >> 4;
+                }
+                dst_pos[i] = b64_table[accum];
+                break;
+            case 2:
+                accum = ((*src_pos) & 0x0F) << 2;
+                src_pos++;
+                if (i < dst_size - 1) {
+                    accum |= ((*src_pos) & 0xC0) >> 6;
+                }
+                dst_pos[i] = b64_table[accum];
+                break;
+            case 3:
+                dst_pos[i] = b64_table[(*src_pos) & 0x3F];
+                src_pos++;
+                break;
+        }
+    }
+    free(src_buf);
+    for (size_t i = 0;  i < pad;  ++i) {
+        dst_pos[dst_size + i] = '=';
+    }
+    return dst;
+}
+
+#define NOT_FOUND (0xFF)
+
+static uint8_t find_index(uint8_t c)
```

Comment:
it might be ok to inline this

Reply:
Or if the compiler is good enough, let the compiler do it.  It would probably do a better job than a human, anyway.

Reply:
Done.  Please verify.

Reply:
thanks :)